design/k8s-job-hook.md

```diff
@@ -0,0 +1,176 @@
+LAKEFS_HOOK_HOOKID - ID of the hook
+LAKEFS_HOOK_REPOSITORYID - ID of the repository
+LAKEFS_HOOK_BRANCHID - ID of the branch
+LAKEFS_HOOK_SOURCEREF - Reference to the source that triggered the event
+LAKEFS_HOOK_COMMITMESSAGE - The message for the commit
+LAKEFS_HOOK_COMMITTER - Name of the committer
+LAKEFS_HOOK_COMMIT_METADATA - Commit metadata (json serialized string)
+```
+
+By default lakeFS will use the following as base definition to schedule a job:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ""
+spec:
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: hook
+        image: ""
+        command: []
+        args: []
+```
+
+Note that the metadata.name, image, command and args will be set by lakeFS.
+Name - unique identifier to identify the specific job execution
+Image, command and args - set based on the hook information
+In case we would like to set a different job specification, we can specify in the lakeFS configuration (lakefs.yaml):
+
+```yaml
+hooks:
+  k8s-job:
+    spec_files:
+      default: /etc/lakefs/job_spec.yaml
+```
+
+Following an example of a `job_spec.yaml`:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: ""
+  namespace: lakefs-hooks
+  labels:
+    app.kubernetes.io/name: post-merge-hook
+    app.kubernetes.io/version: "1.0.0"
+spec:
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: post-merge-hook
+        app.kubernetes.io/version: "1.0.0"
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: hook
+          image: ""
+          command: []
+          args: []
+          env:
+            - name: SPECIAL_ENV
+              value: special_value
+          resources:
+            limits:
+              cpu: "2"
+              memory: 4G
+            requests:
+              cpu: "1"
+              memory: 2G
+```
+
+We can mount the job spec using a configmap having lakeFS use it as the base to the job template used for all k8s-job hooks.
+
+*Limit the end-user image use*
+
+Using the lakeFS configuration, we can specify a list of allowed images that the end-user and use in the k8s-job hook:
+
+```yaml
+hooks:
+  k8s-job:
+    allowed_images:
+      - rclone/rclone:1.57
+      - alpine
+```
+
+Each item list will match the `registry/name:tag` used in the hook.
+If the tag is omitted, any tag will be allowed.
+
+### Execution
+
+lakeFS requests job creation from the cluster.  In case the request fails it will log the job information and fail the hook execution.
+By default the hook waits for the job execution to complete, and based on the job status consider the execution to be successful.
+It will capture the job output into the step output.
+In case we would not like to block the operation (merge/commit) until the job execution is completed, for example on post-* events. We can change the default behavior by setting wait_for_complete to false:
+
+```yaml
+properties:
+  wait_for_complete: false
+```
```

Comment:
`post-*` events are by default non-blocking. They can't block something that already happen. `Blocking` here simply means postponing the response to the client for something that already available to others.

Reply:
reduced the proposal - will not include blocking capability